### PR TITLE
Extract common IonQ optimizer

### PR DESCRIPTION
**Pull request by speller26**

### CODE REVIEW

**Review by PicksPeng — APPROVED** [review]
Looks good.
The Rz gates after the circuits are necessary since there might be measurement-related gates post the complied circuits. For now, we have not implemented them, but when the need comes, it preserves the expected behaviors of the circuits.

File: src/simuq/braket/braket_ionq_provider.py
Comment (licedric) [inline src/simuq/braket/braket_ionq_provider.py:1]: should be braket_ionq_transpiler.py

File: src/simuq/ionq/ionq_api_circuit.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+from simuq.backends.ionq_circuit import IonQCircuit, to_turns
+
+
+class IonQAPICircuit(IonQCircuit):
+    def __init__(self, n_qubits, name, backend="simulator", noise_model=None):
+        super().__init__(n_qubits)
+        # Circuit are stored in turns, accum_phases are stored in rads.
+        self.job = {
+            "lang": "json",
+            "shots": 4096,
```
Comment (licedric) [inline src/simuq/ionq/ionq_api_circuit.py:12]: Seems like a dangerously high shot count. This is ultimately overwritten later on but I'd make this smaller for safety

Comment (licedric) [reply src/simuq/ionq/ionq_api_circuit.py:12]: In retrospect, this probably also means the default shot count in the IonQProvider should be adjusted downward

File: src/simuq/ionq/ionq_api_circuit.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+                q0, q1 = gate["targets"]
+                phi0, phi1 = np.array(gate["phases"]) * 2 * np.pi
+                angle = gate["angle"] * 2 * np.pi
+                new_circ.ms(q0, q1, phi0, phi1, angle)
+            else:
+                raise Exception("Unknown gate:", gate["gate"])
+
+        for q in range(n):
+            new_circ._add_unitary(q, unitaries[q])
+            new_circ.rz(q, -self._accum_phases[q])  # TODO: is this really necessary?
```
Comment (licedric) [inline src/simuq/ionq/ionq_api_circuit.py:94]: This comment can be removed, I suppose

**Comment by licedric** [comment]
Other than the comments above, LGTM

File: src/simuq/ionq/ionq_api_circuit.py (hunk 1 above)
Comment (licedric) [reply src/simuq/ionq/ionq_api_circuit.py]: Maybe shots = 0 here and 10 in the IonQProvider? @PicksPeng for visibility

**Comment by licedric** [comment]
Can you also update the shot counts for the IonQProvider in the notebooks? Thanks!

**Pull request merged**